language/bytecode_verifier/src/type_memory_safety.rs

```diff
@@ -237,9 +237,17 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
             }
 
             Bytecode::Ret => {
+                for idx in 0..self.locals_signature_view.len() {
+                    let is_reference = state.is_available(idx as LocalIndex)
+                        && state.local(idx as LocalIndex).is_reference();
+                    if is_reference {
+                        state.destroy_local(idx as LocalIndex)
+                    }
+                }
                 if !state.is_safe_to_destroy() {
                     return Err(VMStaticViolation::RetUnsafeToDestroyError(offset));
                 }
+
```

Comment:
Would it be sufficient to check that every resource local is unavailable instead?

I was looking at the logic for `destroy_nonce` (called by `destroy_local` on a ref) and it seems fairly complex. I was thinking that it might be simpler/safer to do the check suggested above if that's what this boils down to. But I couldn't determine if additional checks like "only references to formals can be returned" are also performed here.

Reply:
I did not understand the alternative @sblackshear is proposing.  If there are outstanding references to locals, those locals would be available.  In this situation, it would be wrong to reject the program.

Reply:
Otherwise, the change looks fine to me.

Reply:
Ah, sorry for being unclear. My question is: what would go wrong if you replaced lines 240-250 with something like

```
for idx in 0..self.locals_signature_view.len() {
  let is_available_resource = state.is_available(idx as LocalIndex) && state.local(idx as LocalIndex).is_resource();
  if is_available_resource {
    return Err(VMStaticViolation::RetUnsafeToDestroyError(offset));
  }
}
```

Reply:
For a program with no resources, the code you proposed will do nothing.  Do you think some checks are warranted even for a program with no resources?  

p.s. I am trying out the Socratic method of answering a question with another question :-)

Reply:
I think the relevant things to check for a `Ret` instruction are:

(1) No local variable holds a resource value
(2) The # of values on the stack/their types match the return type of the current procedure
(3) Any reference that is being returned is either a formal of reference type or borrowed from a formal of reference type.

(2) and (3) should still be checked on a program with no resources. It looks like (2) is checked by the code from 251-261. This leaves (3). So I suppose my new question is:

- Did I get (1) - (3) right?
- Can (3) be checked in a simpler way than in the current scheme (e.g., check that `operand`s of type reference borrow from a formal at line 258)?

Reply:
(1)-(3) look correct to me.

The current scheme checks (3) indirectly.  It checks that there are no borrows on local variables at the point control transfers back to the caller.  This check is stronger than (3), i.e., if this check succeeds then (3) is guaranteed.

Reply:
I guess what my series of questions is getting at is: would it be better to check (1) - (3) directly instead of checking stronger conditions that indirectly imply (1) -(3)? I don't think that was realistic before because of the restrictions on releasing references, but I think it is realistic (and perhaps preferable) now.

Reply:
By releasing all of the references stored in locals first, (1) and (3) are directly checked by `is_safe_to_destroy`. There is no stronger property being checked 

(3) cannot be checked simpler in the current borrow checker, as we do not have the mapping to make that check possible. So by releasing the references in locals, we have "A local is borrowed" iff "A reference rooted in a local is being returned"

Reply:
I think though that we should likely move the check of `is_safe_to_destroy` to after the checking of the signature. It makes more sense to me to first check the `Ret` instruction itself, and then check the surrounding environment